### PR TITLE
[MAINTENANCE] Move "Domain" to "great_expectations/core" to avoid circular imports; also add MetricConfiguration tests; and other clean up.

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Move `Domain` class to “great_expectations/core” directory in order to avoid circular imports; moreover, the `Domain` class has become a truly core entity (and will be even more so going forward).
* Add `MetricConfiguration` tests.
* Add `repr()` methods to `ValidationGraph` and `MetricEdge` classes.
* Clean up type hints in several modules.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-1347/GREAT-181/GREAT-1426
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/validator/metric_configuration.py [hunk 1]
```diff
@@ -65,6 +67,30 @@ def metric_dependencies(self) -> IDDict:
...
+            "column_A" in self._metric_domain_kwargs
+            and "column_B" in self._metric_domain_kwargs
+        ):
+            return MetricDomainTypes.COLUMN_PAIR
+
+        if "column_list" in self._metric_domain_kwargs:
+            return MetricDomainTypes.MULTICOLUMN
+
+        # TODO: <Alex>Determining "domain_type" of "MetricConfiguration" using heuristics defaults to "TABLE".</Alex>
+        return MetricDomainTypes.TABLE
```
Comment (NathanFarmer) [inline great_expectations/validator/metric_configuration.py:92]: Does this have anything to do with moving `Domain` or is this new functionality?

**Review by NathanFarmer — COMMENTED** [review]
This seems like a great idea beyond the circular import issues since we seem to have a need to incorporate `Domain` formally into more areas of the codebase. I just have one question about the new methods.

File: great_expectations/validator/metric_configuration.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/validator/metric_configuration.py:92]: @NathanFarmer This is a "slightly" new functionality (as we have something similar in `ExpectationConfiguration`.  Thanks.

**Review by NathanFarmer — APPROVED** [review]
LGTM!

**Pull request merged**